Add DAISY descriptor for wide-baseline / keypoints.

cbalint13 opened the pull request:
This PR add DAISY [1] descriptor specialized for dense stereo problems. It is very fast for  wide-baseline [2] but can be used just as normal keypoint descriptor generated by some detector, with quite superior performance among other descriptors.

Some more infos:

* Implementation here is properly cited in respect with authors.
* Implementation is derived from latest author's original BSD licensed code available in public.
* Implementation booth follows opencv's coding standard and author original code performance.
* Implementation behave 1:1 numerically and by speed with original.
* Implementation pass well QA with very good rotation invariance properties.
* Implementation pass poorly QA for scale invariance (this is expected !)
* Implementation pass with excellent results real SfM scenarios for dense stereo-matching.
* Implementation is flexible its parameters are well documented and tested.

[1] http://cvlab.epfl.ch/software/daisy/
[2] http://cvlab.epfl.ch/research/surface/emvs/

A cross PR for QA: https://github.com/Itseez/opencv_extra/pull/254

cbalint13 commented:
All passed.

* Current opencv_contrib buildbot have perf/test disabled.
* Squashed all into https://github.com/cbalint13/opencv_contrib/commit/96f1e2566096e3c95098b95273aef8b271a36070

StevenPuttemans commented:
Nice!

vpisarev commented:
@cbalint13, thank you a lot for the contribution! We planned to put DAISY into OpenCV for a long time, glad you finally working on it.

I will look into details later, but have a few immediate comments/suggestion regarding the API.
DAISY could be used as sparse descriptor, of course, but its main role is dense descriptor. I think, it would be very, very inefficient to construct a huge array of keypoints and then get descriptors for all of them. More efficient approach would be to add a dedicated method to compute descriptors for each pixel in a ROI of image. Why ROI? It's convenient for parallel processing (by tiles), it's enough information to process to compensate for overhead and it's less space for output descriptors vs the whole image case (but user can always pass ROI covering the whole image if needed).

Of course, before that the image should be passed and preliminary calculations should be made. That is, the following method could be added to DAISY:

class DAISY ... {
...
// stores the collection of filtered images, from which the descriptors can be extracted.
virtual void setImage(InputArray img, <probably some extra params>) = 0;
// stores descriptors for the specified ROI into 2D array (roi.width*roi.height x size_of_DAISY_descriptor). // note the constness, allowing parallel execution.
virtual void compute(Rect roi, OutputArray descriptors, <some more params, perhaps>) const = 0;
};

what do you think?

cbalint13 commented:
@vpisarev ,

Lets clarify DAISY behaviour at this moment of implementation, and its main work-mode flags:

1. DAISY::COMP_FULL when keypoints are ignored and must be _blank_ (there is an assert to check this in this very mode) all pixels are computed, thus in the end the descriptors resulted for Image(x,y) can be accessed by something like this: descriptors.ptr<float>( y*descriptorSize() + x ). I took _special_ care in this dense mode to be as fast as original code and avoid any kind of .clone() or other things, i just reference author's original container into final cv::Mat(), see here: https://github.com/Itseez/opencv_contrib/pull/229/files#diff-453dc15e0b4d66e076bc9e89749e3a13R1970 In this mode DAISY is blazing fast.
  Your statement "very, very inefficient to construct a huge array of keypoints" is not true, i asserted into code,  in this very mode keypoints to be always supplied as _empty_.

2. DAISY::ONLY_KEYS , the classical, when keypoints can be supplied and descriptor is computed _only_ for each keypoint: see  https://github.com/Itseez/opencv_contrib/pull/229/files#diff-453dc15e0b4d66e076bc9e89749e3a13R1980I do not describe anyomore this mode, it is clear its scope. I am impressed about rotation invariance properties in such usage mode, it surpassed my expectations, (never used daisy before in image matching scope).

Lets further improve given your nice suggestions:

1. Agree with "virtual void compute(Rect roi, OutputArray descriptors, ) const = 0;" in DAISY::COMP_FULL mode. Will implement it right now ! Its a brilliant alternative idea for _blank_ keypoints which is ment only for whole image. Obviously  will do some "pragma parallel" for the rectangle collections, only if will not conflict/race with DAISY internals parallel statements (will check the improvements and see if there are races / impacts).
  Technically with ROI we might encounter problems in the code, i am afraid that:
  a) Will have to call daisy for each subimage, then finally merge all descriptor arrays for those regions into single one array. Need to see if array can be reused as buffer at all to avoid final merge and multiple buffers This is most the feasible solution for this. 
  b) Can teach daisy internals to compute only for ROI areas, not so easy really have to *master* original daisy code ! But lets see the code first and then talk, perhaps it is not so hard.
  For booth a. or b. descriptors.ptr<float>( y*descriptorSize() + x ) would return zeros if user is out of ROI. Or, as alternative,  plan to expose public get_descriptor(x,y) that is aware of ROI with hardcoded assert if out of region.

2. One little perhaps uncovered and demanded thing  would be to add a third mode called DAISY::ONLY_QUERY which behave like ONLY_KEYS, but no keys should be supplied at all (will assert this), only some sparse precomputations are done by daisy on whole image and cached in the object (for little speedup) but would allow on-demand computation for any given and queried (x,y) and even using an "r" rotation (x,y,r) factor too. Will expose a public class like get_descriptor(x,y,r). In contrast please notice that COMP_FULL does not allow an extra "r" (due to its speedy mode) , thus why this mode is demanded.

3. A more clear documentation with all above.

4. I plan add a _very_ nice example (promise) of dense-stereo problem (no MRF !) once this PR is merged.

 Will do so, be right back asap. I was using daisy before in dense-basline problems, i really can tell many things about original implementations and authors intentions. In meanwhile i (with polite) suggest any reviewer to read "usage" from daisy original tarball.

vpisarev commented:
@cbalint13,
thank you for the prompt reply. Unfortunately, github does not allow to answer with citation. So, I just have my answers split into several items that we discuss.

* ah, ok, I now see that there is special mode for this. Yet, I'd prefer to do it in less hacky way, but rather explicitly using the extra method.

* It's good to see that DAISY demonstrates good performance as "sparse" descriptors. BTW, I wondered how they treat the keypoint "scale" and apparently, it's just ignored. I think, this should be mentioned somewhere in the description.

* On parallelizing DAISY. I think, it's not high priority for now, but replacing "#omp pragma parallel for with" with cv::parallel_for_ would be useful at some point, maybe later.

* On the problems with ROI. I shall emphasize that the whole image is first passed (using setImage() method) and preprocessed and then descriptors are extracted only for the ROI - this compute() may be called multiple times after a single setImage(). I think, it should not be difficult to do - modify compute_descriptors:
       1. add "Rect roi" parameter, pass m_descriptors as the output parameter (e.g. "out_descriptors") instead of accessing class member. Call out_descriptors.create(roi.height*roi.width, <DAISY_SIZE>, CV_32F);
       2. instead of the loop over 0..h x 0..w do the loop over roi.y..roi.y+roi.height x roi.x..roi.x+roi.width.
       3. in out_descriptors use the row index "(y - roi.y) * roi.width + (roi. x - roi.x)" to store the computed descriptor. 

   So, on the problem a) - usually, there is no need to keep all the computed descriptors. If you do stereo for a rectified stereo pair, you just need a stripe, even a line at once, which you store to temporary buffer. As long as you've computed the disparity, the descriptors are not needed anymore.

   On the problem b) - the solution is above. 

   The mode 3), as I understood it, is a partial case of both the compute method that I suggested (when (x,y) are integers) - use Rect(x, y, 1, 1)) or the existing compute when there is just one keypoint. I believe, user code can be just adjusted to extract several descriptors at once, this is usually always good idea and helps to decrease the overhead. So, do not worry about ONLY_QUERY.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+ */
+
+#include "precomp.hpp"
+#include "opencv2/imgproc/imgproc_c.h"
```
could you please modify the code to use the modern C++ API and thus do not use imgproc_c.h?

cbalint13 replied:
DONE.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+    int m_w;
+
+    // stores the descriptors : its size is [ m_w * m_h * m_descriptor_size ].
+    float* m_dense_descriptors;
```
m_dense_descriptors should not be stored, otherwise the method that extracts descriptors for ROI can not be made "const"

cbalint13 replied:
* m_dense_descriptor is now Mat type. I did extractors as const. [DONE]

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+
+    // stores the layered gradients in successively smoothed form: layer[n] =
+    // m_gradient_layers * gaussian( sigma_n ); n>= 1; layer[0] is the layered_gradient
+    float* m_smoothed_gradient_layers;
```
use Mat or vector&ltMat&gt; or vector&lt;float&gt; do not use plain arrays and manual memory management

cbalint13 replied:
No more alloc() / dealloc() or new [] ! Only Mat types, and where was necessary  accessed data via mat.ptr() schema. [DONE]

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+    int m_orientation_resolution;
+
+    // hold the scales of the pixels
+    float* m_scale_map;
```
same

cbalint13 replied:
m_smoothed_gradient_layers is now Mat. It's the most important matrix of floats in daisy. Unfortunate is that is N dimensions (i think N > 5) so in many place i was able to use it as .ptr() reference without undersanding its internal store system. Speed was not affected by conversion at all. [DONE]

cbalint13 replied:
m_scale_map too. its now 2d Mat with clear undrstandings whats inside. [DONE]

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+    float* m_scale_map;
+
+    // holds the orientaitons of the pixels
+    int* m_orientation_map;
```
same

cbalint13 replied:
m_smoothed_gradient_layers, m_scale_map ,m_orientation_map are Mat now (2d, 1d dimensions). I was able to use nice .row() accessors in some cases. In fact we now have only Mat matrices. [DONE]

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+    int* m_orientation_map;
+
+    // Holds the oriented coordinates (y,x) of the grid points of the region.
+    double** m_oriented_grid_points;
```
same. Also, please change "double" to "float" where possible. I think, the case of DAISY everything could be made "float" without any accuracy problems. float's are several times faster than double's on ARM. On some platforms float's are available in hardware, but double's are not - then the difference becomes catastrophic.

cbalint13 replied:
m_oriented_grid_points is now 2d Mat of CV_64F. I believe it is important to stay CV_64F, however it is small one. This double pointer array was ugly, i agree. [DONE]

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+
+    // holds the gaussian sigmas for radius quantizations for an incremental
+    // application
+    double* m_cube_sigmas;
```
same

cbalint13 replied:
m_cube_sigmas too, it is a CV_64F (small one) and is important for precision sampling. [DONE]

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+    void set_normalization( int nrm_type ) { m_nrm_type = nrm_type; }
+
+    // applies one of the normalizations (partial,full,sift) to the desciptors.
+    void normalize_descriptors( int nrm_type = DAISY::NRM_NONE );
```
such methods should include necessary input/output parameters, like "Mat& descriptors"

cbalint13 replied:
These are private internals, will never expose such things for public. Did you really want to pass through them Mat &descriptors ? [PENDING]

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+
+    inline void conv_horizontal( float* image, int h, int w, float* kernel, int ksize )
+    {
+      CvMat cvI; cvInitMatHeader(&cvI, h, w, CV_32FC1, (float*)image);
```
please, use cv::filter2D (or better more efficient cv::sepFilter2D to convolve ((image * kernelX) * kernelY) in one call

cbalint13 replied:
[DONE]
* I used now filter2D twice for H,V steps. Was not able to use the nice getGaussianKernel() because it differs in results at 10e-6 and daisy produced different results ! Also was not able to use sepFilter2D(), but filter2D worked perfect using BORDER_REPLIACATE and not default one. 
* Converted to opencv, and we still produce 1:1 at bit level same numerical results.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+     * DAISY utilities
+     */
+
+    template<class T>
```
can it be replaced with cv::Rect?

cbalint13 replied:
[DONE], those templates are now 100% gone.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+
+    // computes the l2norm of an array: [ sum_i( [a(i)]^2 ) ]^0.5
+    template<class T> inline
+    float l2norm( T* a, int sz)
```
opencv has such inline functions.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+
+    // allocates a memory of size sz and returns a pointer to the array
+    template<class T> inline
+    T* allocate(const int sz)
```
manual memory allocation should be replaced with std::vector or cv::Mat

cbalint13 replied:
[DONE] no more alloc() / dealloc() or new[]. We have only nice Mat.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+
+
+    template<typename T> inline
+    void convolve_sym_( T* image, int h, int w, T* kernel, int ksize )
```
it can be done with cv::sepFilter2D

cbalint13 replied:
[DONE], commented already (upper) on the conv issue, that template too is now gone.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+    }
+
+    // returns an array filled with zeroes.
+    template<class T> inline
```
use Mat::zeros() for such things, or std::vector::asign().

cbalint13 replied:
[DONE] we have now Mat(r,c,CV_32F,Scalar(0)) or mat.setTo(Scalar(0)). I did not used std::vector in the code, Mat is much more flexible and can be casted /converted / re-dimed with much more ease.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,2057 @@
+    // to be an odd number.  if in_place=true, then T1 must be equal
+    // to T2 naturally.
+    template<class T1, class T2> inline
+    T1* blur_gaussian_2d( T2* array, int rn, int cn, float sigma, int kernel_size = 0, bool in_place = false )
```
let's replace this with cv::GaussianBlur

cbalint13 replied:
[DONE] but with 2 x filter2D to be 1:1 bitwise level with same results as original.

cbalint13 commented:
@vpisarev ,

 Agreed with your initial request and dropped the DAISY::COMP_FULL / DAISY::ONLY_KEYS (quite bad to maintain such, and needed a lot of stupid assert to warn users thei are in wrong sittuations) in the favour of elegant overloaded compute()'s. Only one purpose compute() functions which are much better, can be re-entrant and re-useable in place without calling whole constructor/destructor once again (as expected). Exposed a set_image() option for those who want to fast re-iterate and short-circuit compute() call things for a new image and did lot of API scenario tests. Now we have:

1. virtual void compute( InputArray image, std::vector<KeyPoint>& keypoints, OutputArray descriptors ) = 0; // The classic one, no need to comment.

2. virtual void compute( InputArray image, Rect roi, OutputArray descriptors ) = 0; // I agreed with you to have a ROI based one, i think is still a clever idea.

3. virtual void compute( InputArray image, OutputArray descriptors ) = 0; // An this is the full mode one.

*  Also exposed "get_descriptor() const" variants for those who want  further "very fast" query and sample points, and want short-circuit a bit the procedure to fill a keypoints list and call compute() again. These get_descriptor are not for the 1,2,3 mode, they are for the single-key mode (as per original daisy does).

* Regarding const-ness for compute() i think is not feasible , daisy already use internally parallel loops so will not gain any additional speed calling multiple images at once. Plus there is a delicate internal workspace memory management that should be in care. If still insist i can do the const thing, but believe me we will gain nothing.

* Fixed memory handle schema in case user re-entrer into compute() with new image / keypoint, or use set_image() scenarios.

-------------------------
Further more thank you a lot for the deep code review and suggestions ! My intention was in that direction too, but first i had to settle the API down , do more test with 1:1 untouched original code (its done at this moment) and only after then try to optimize and change the code for more opencv style. I will do _ALL_ of your suggested changes here and test&assure that descriptors generated are still the same and behave the same.

cbalint13 commented:
@vpisarev ,

 Work is in progress. I did all your requests upper and commented (for myself to know i did not miss something), plus many more optimizations to the code. It got quite "slim" one after these changes. Later today i push the new code for review, now i am checking through:

1) Should generate 1:1 same code as original [done]
2) Should pass QA as original [done]
3) Should pass real SfM as keypoint descriptor [done]
4) Should pass a real rense-basline densification and same amount of points should be obtained [pending]


Be back,

cbalint13 commented:
@vpisarev ,

All requests done. Will fix win64 warns too later. In the meanwhile you can review once again the code, here is the squashed one: https://github.com/cbalint13/opencv_contrib/commit/fe38c9ef469d079568cd287a1f9c1771fda50bc5

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,1754 @@
+
+   Mat cvO;
+   // smooth the data matrix
+   filter2D( data, cvO, CV_32F, Kernel, Point( -1, -1 ), 0, BORDER_REPLICATE );
```
could we replace it with GaussianBlur() or sepFilter2D()?

cbalint13 replied:
* As i mention it i was nto able to replace the two filter2D() with single sepFilter2D() without numeric impact in final descriptor. If doing it produced descriptors will not be numericaly 1:1 as original daisy.

 If numerical result slightly differ and this is acceptable here, then  yes i will replace.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,1754 @@
+
+   Mat dx(1, data_size, CV_32F);
+   Mat dy(1, data_size, CV_32F);
+   gradient( cvI, data.rows, data.cols, dy, dx );
```
could we replace it with 2 cv::Sobel() calls?

cbalint13 replied:
I'll try. Only if you confirm that would be acceptable to have slight different numeric results (but obviously same quality) as original daisy.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,1754 @@
+   Mat dx(1, data_size, CV_32F);
+   Mat dy(1, data_size, CV_32F);
+
+   gradient(cvO, data.rows, data.cols, dy, dx);
```
could we replace it with two cv::Sobel() calls?

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,1754 @@
+   Mat Kernel(1, 5, CV_32F, (float*) kernel);
+
+   filter2D( cvI, cvI, CV_32F, Kernel, Point( -1, -1 ), 0, BORDER_REPLICATE );
+   filter2D( cvI, cvI, CV_32F, Kernel.t(), Point( -1, -1 ), 0, BORDER_REPLICATE );
```
GaussianBlur? or sepFilter2D?

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,1754 @@
+    m_descriptor_size = 0;
+    m_grid_point_number = 0;
+
+    m_grid_points.release();
```
the .release() calls are not necessary - all the matrices are empty at the beginning.

vpisarev commented on modules/xfeatures2d/src/daisy.cpp:
```diff
@@ -0,0 +1,1754 @@
+
+    // iterate over keypoints
+    // and fill computed descriptors
+    if ( H.empty() )
```
looks like there is almost complete code duplication, whereas the only difference (if I understood correctly) is point_transform_via_homography() call. If you pass NULL pointer as "double* H" and check it before calling point_transform_via_homography() - that will be very efficient, because the condition will be predicted perfectly by CPU and you loose may be just a few CPU clocks per descriptor (and computing a single DAISY descriptor takes 100's of clocks)

vpisarev commented:
@cbalint13, thank you for working on it! I have a few more comments, regarding elimination of custom gaussian blur & sobel. Switching to standard ones will reduce the further optimization efforts and simplifies long-term support. Also, as I see that the computed descriptors are still kept in the class. It means that you can not extract descriptors from different ROIs in the same image from parallel threads and thus you can not use DAISY in parallel stereo algorithm without obtaining all the descriptors at once (which may take a lot of memory!). It can be addressed later, I just wonder whether you were going to make this change before integration or not.

cbalint13 commented:
@vpisarev 

1. Will try do the proposed replacements. Until now my goal was 1:1 same numerical output as original. If will be acceptable (?!) we will have slightly different numerical output compared to the original DAISY but will try to have same QA , with some chance perhaps slightly better, i wont be surprised :)

2. Will remove m_dense_descriptor out form the class ! I think it is 100% possible to do it.

be back,

vpisarev commented:
1. DAISY uses floating-point arithmetics. Therefore, you cannot assume to get bit-exact results as in the original code. On different OSes, different compilers, different CPUs with different IEEE754 implementations you can get slightly different descriptors. Besides, as I see, there are many optimization opportunities in this code, so we'll likely diverge from the original code even more. Therefore, I do not see a reason not to do it now :)

2. ok, thank you! I think, this feature (const'ness of the method that computes descriptors for a part of the image) will be really appreciated!

vpisarev commented:
excellent! :+1:

cbalint13 commented:
@vpisarev ,

Wanted to tell you:

- Converted all arrays into nice Mat arrays and handle them properly to not loose performance.
- Used GaussianBlur, Sobel for more opencv style handling, difference is only at ~10^-7 level.
- We are still in QA for booth opencv's QA and real SfM densification problem, i always checked upon after any small modifications to the code.
- Most important:

 * Could _not_ do "const" for compute(_with_roi), it does *not* make sense at all because DAISY works better in such way:
     1. m_smoothed_gradient_layers, m_image and many more are _pre_computed_ and cached in  the class.
     2. You can ask for full descriptor computation and then destroy your object but most of the time ~95% people will  re-invoke in iterative way exposed  GetDescriptor() const variants with various rotation or various x,y during their agressive but selective baseline scan until 3D model converge, while ptr<DAISY> still caches important pre-computations for each GetDescriptor() calls ! It is hard to tell how many queries will be done, it depends how 3d model converge. Exactly how original daisy was designed and ment for.
     3. If people wants multiple images that construct for themselves multiple ptr<DAISY> classes each with its own caches of point no. 1 and they just will scan using GetDescriptor() const variant within their multiple threads.
    4. Doing compute() const is lousy, will lost all our internal per class caches and all the time daisy will recompute that with precious amount of time, point 2 here is much smarter approach.

- It is trivial to show a sample/tutorial, and i will work one such to show how to use it _efficiently_.
- We are twice slower than original and very little chance to speed up due to the fact raw original OpenMP is more powerful then cv::parallel_for. I tried with untouched original pieces of daisy within cv::parallelel_for and never had the same speed, but the code is very elegant now indeed, and easy to understand the objects there.
- My intuition says that additional speed might be gained when will further use GetDescriptor() const within multiple threads on top of internal cv::xfeat::daisy parallelism but will point it out  when tutorial will be done.

vpisarev commented:
@cbalint13, thank you!
1) do not worry about speed for now, there are many places in the code where we can compensate for the speed. For example, parallel_for_ has the optional "nstripes" parameter, which you can play with and probably get all the performance back.
2) constness of compute can be provided, but with 2-stage approach that I tried to describe earlier. First, user calls non-constant setImage(image) method, where all the maps are computed for the whole image. And then "compute(Rect roi, OutputArray descritptors) const" method is called without the input image and simply uses the precomputed maps to extract descriptors for the specified ROI.

cbalint13 commented:
@vpisarev,

Vadim,

   I started work on a nice tutorial to add for opencv_contrib, it will be a real post-SfM densification of ~10 images (fountain scene) producing a very dense .PLY model. Will still keep account of what you proposed, will not let those wishes to "just go". I plan to outline the various modes that DAISY can be used. I worked hard for conversion into the nice Mat structures (e.g. used much more optimal ND Mat where was apropiate, and re-casted into sub 2d,3d array views where was needed),  i needed to fully understood booth author papers and the original code and the data structures layout that where quite "entangled". Hope people will like it. I agree with the performance constraints, in fact i like it now much more in this form.

 Thank you **very** much for helping me on it !

vpisarev commented:
@cbalint13, from my side I would like to thank you for this excellent contribution! In fact, I did not expect that you would make this much refactoring of the code :) and really glad you did it!

edgarriba commented:
@cbalint13 are you working on this post-Sfm densification with DAISY? I have some lines of code for this following the authors papers.

cbalint13 commented:
@edgarriba,

- Yes. I tried to replicate EMVS (post densification) from author's paper. I also have some lines of code :). Since code is messy would share in private with all the explanations. My goal is to publish it, but not in this state.

Sent you message by e-mail.
